WordPress/gutenberg
# Popover: lock the __experimentalPopoverPositionToPlacement function

## What?


Part of #47196
**Requires #47229 to be merged first**

In an effort to reduce the amount of experimental APIs introduced to the repo, this PR uses [the experimental `lock`/`unlock` APIs](https://github.com/WordPress/gutenberg/pull/46131) to lock the `__experimentalPopoverPositionToPlacement` function from the `@wordpress/components` package, renaming it to `__experimentalPopoverLegacyPositionToPlacement `.

## Why?


I had initially decided to stabilise this APIs, but following the conversations in this PR I later changed the approach and went for `lock` / `unlock`. The whole point here is to prevent the `__experimentalPopoverPositionToPlacement` from being released to core and therefore losing its "experimental" nature.

Also, I decided not to introduce any soft deprecations and directly rename the export because it doesn't look like [there any usages according to WPDirectory](https://wpdirectory.net/search/01GQSVMVKQDRA8QV605BVQM858).

## How?


- Removed the original export from `@wordpress/components`, and instead added the function under the `experiments` object (which is locked)
- Refactored the consumers (`URLPopover`) to use the `unlock` APIs to retrieve the same function

## Testing Instructions





- Make sure project build
- Make sure `Popover` Storybook example work as expected
- Make sure that the `URLPopover` is positioned correctly in the editor (as on `trunk`)
- Make sure that there aren't any leftover mentions of the `__experimentalPopoverPositionToPlacement` prop

<details>

<summary> Detailed instructions on how to test the `URLPopover` component</summary>

1. Apply this diff in order to use the legacy `position` prop, which should trigger the unlocked function

```diff
diff --git a/packages/block-editor/src/components/url-popover/image-url-input-ui.js b/packages/block-editor/src/components/url-popover/image-url-input-ui.js
index 57084ae76c..aa51c4ba58 100644
--- a/packages/block-editor/src/components/url-popover/image-url-input-ui.js
+++ b/packages/block-editor/src/components/url-popover/image-url-input-ui.js
@@ -252,6 +252,7 @@ const ImageURLInputUI = ( {
 			/>
 			{ isOpen && (
 				<URLPopover
+					position="top"
 					anchor={ popoverAnchor }
 					onFocusOutside={ onFocusOutside() }
 					onClose={ closeLinkUI }
```

2. In the post editor, insert an image block and add an image to it
3. Select the image, and in the block toolbar click on the "Insert link" button
4. Make sure that the resulting `URLPopover` renders on top of the block toolbar, and that a warning is printed to the browser's console (since `position` is deprecated)
